### ISSUE TITLE
Improve Search to Consider Popularity/Newness

### ISSUE THREAD

**Issue by ebollens** [universityofcalifornia/NeXt#31]
Currently, search simply returns the best match. This should be improved so that it returns not just based on best match, but also factors in popularity for ideas (the number of "likes") and new-ness for projects and ideas.

**Comment by ebollens** [comment]
Made some progress on this today. Definitely not PR-worthy yet (needs tuning, additional scaling factors, more research into scaling functions, etc.), but I wanted to share a bit about my approach to solicit feedback...

So, before I address the front page, I'm starting with the `/search` page since it's already hooked into Elasticsearch for its results. The current query for it is:

```ruby
    search_body['query'] = {
      'bool' => {
        'must' => [
          {
            'multi_match' => {
              'query' => params[:query],
              'type' => 'phrase_prefix',
              'fields' => ['_all']
            }
          }
        ]
      }
    }
```

This returns the following when I search for "Ruby" under "Projects":

Project | Score
------------ | -------------
WebBlocks | 0.210471
NeXt Ideation Platform | 0.1999431

This is non-ideal because...

* NeXt is newer
* NeXt has more votes

But that's happening now? Basically, both have Ruby as a necessary competency, which is why they match, and the difference in scoring comes from the comparison between the matched search space and the total available search space.

A basic primer on [relevancy](https://www.elastic.co/guide/en/elasticsearch/guide/current/relevance-intro.html) is a good place to start. Going a step further, after some experimentation, it appears the right approach here is going to be using [`function_score`](https://www.elastic.co/guide/en/elasticsearch/reference/2.0/query-dsl-function-score-query.html) as a wrapper around the search query:

```ruby
search_body['query'] = {
      "function_score": {
        "query": {
          'bool' => {
            'must' => [
              {
                'multi_match' => {
                  'query' => params[:query],
                  'type' => 'phrase_prefix',
                  'fields' => ['_all']
                }
              }
            ]
          }
        },
        "functions": [
          # scoring f(x)
        ]
      }
    }
``` 

The big challenge here is figuring out what replaces `# scoring f(x)`.

Here's what I've got so far, focused on scaling the scores for dates based on a curve, where that curve is an exponential function centered on today with a 6 month scale (probably make this vary over time based on how far apart the earliest and latest items are), an offset of 1 month (so no reduction in score occurs for new items) and a decay which specifies the adjustment by deviation (with no real reason for the number chosen thus far, but rather to be tuned based on other scoring functions I also believe we need to add):

```ruby
  search_body['query'] = {
      "function_score": {
        "query": {
          'bool' => {
            'must' => [
              {
                'multi_match' => {
                  'query' => params[:query],
                  'type' => 'phrase_prefix',
                  'fields' => ['_all']
                }
              }
            ]
          }
        },
        "functions": [
          {
            "exp": {
              "created_at": {
                "origin": Time.now.strftime('%Y-%m-%dT%H:%M:%S%:z'), #"2015-09-27T23:50:00",
                "scale": "180d",
                "offset": "30d",
                "decay": 0.8
              }
            }
          }
        ]
      }
    }
```

The result of this:

Project | Score
------------ | -------------
NeXt Ideation Platform | 0.13605121
WebBlocks | 0.073619545

This is only one example of one small subset of what we need to accomplish, but it demonstrates the effectiveness of this approach. Next up will be adding some additional parameters (such as for popularity) and then tuning the scale and decay factors to get a reasonable balance in the way weighting occurs.

**Comment by ebollens** [comment]
So, after a lot of experimentation, I've got the search query now considering:

* Search term likeness
* Preference towards ideas/projects over people
* Preference towards newness (especially < 30 days old)
* Preference towards higher votes

As an example, consider:

Project | Creation Date | Votes
------------ | ------------- | -------
NeXt Ideation Platform | Oct 2014 | 7
Uberize IT resources | Aug 2015 | 10
WebBlocks | May 2013 | 3

Using the old methodology, we end up with something like:

Project | Creation Date | Votes | Weight
------------ | ------------- | ------- | -----
NeXt Ideation Platform | Oct 2014 | 7 | 0.39
Uberize IT resources | Aug 2015 | 10 | 0.21
WebBlocks | May 2013 | 3 | 0.25

If we penalize oldness more by changing the decay to `0.6`:

Project | Creation Date | Votes | Weight
------------ | ------------- | ------- | ------
NeXt Ideation Platform | Oct 2014 | 7 | 0.27
Uberize IT resources | Aug 2015 | 10 | 0.20
WebBlocks | May 2013 | 3 | 0.10

This significantly penalizes WebBlocks, as it's so much older.

Next up, we can change the offset to `90d`, which completely removes the penalty on things proposed in the last 90 days:

Project | Creation Date | Votes | Weight
------------ | ------------- | ------- | ---
NeXt Ideation Platform | Oct 2014 | 7 | 0.27
Uberize IT resources | Aug 2015 | 10 | 0.22
WebBlocks | May 2013 | 3 | 0.10

This leads to the following query expression:

```ruby
          {
            "exp": {
              "created_at": {
                "origin": Time.now.strftime('%Y-%m-%dT%H:%M:%S%:z'), #"2015-09-27T23:50:00",
                "scale": "270d",
                "offset": "90d",
                "decay": 0.6
              }
            }
          }
```

Next up was figuring out how to penalize user results. Basically, as they stand now, because users have less details in them then projects and ideas, they get brought to the top of the result list. 

I tried some stuff with scoring functions written in Groovy - first dynamically, then statically - but it felt very clunky and super arbitrary. So, after going down this path for a while, I abandoned it.

The next approach I tried relied on the fact that ideas and projects both have `votes` but users do not. What this means is that I could introduce some sort of a scaling factor applied off of votes, and make the default value when the votes don't apply to be 1. After some experimentation and tuning, this seemed to have worked pretty well:

```ruby
          {
            "field_value_factor": {
              "field": "votes",
              "factor": 4,
              "modifier": "sqrt",
              "missing": 0.25
            }
          }
```

This ends up giving:

Project | Creation Date | Votes | Weight
------------ | ------------- | ------- | ---
Uberize IT resources | Aug 2015 | 10 | 1.39
NeXt Ideation Platform | Oct 2014 | 7 | 1.28
WebBlocks | May 2013 | 3 | 0.18

And every so nicely, users created recently slip into the middle, not prioritized as highly as the recent projects or ideas, with their votes, but higher than ancient ones...

<img width="1234" alt="screen shot 2015-10-10 at 11 39 11 am" src="https://cloud.githubusercontent.com/assets/761092/10412719/90937dce-6f43-11e5-992e-51c8013a94d2.png">

**One catch to this for @stevenolen to consider is that it requires Elasticsearch 1.7.1.** This is because of the `field_value_factor` scoring function, which does not support `missing` until then.
